### PR TITLE
Import counts

### PR DESCRIPTION
**Pull request by mumichae**
Changes:

+ Large structural changes to DROP implementation. New config classes created, config parsing separated for different modules
+ Removed snakemake subworkflows. All modules steps are run on the same level, so full snakemake functionality should be available for rerunning the pipeline
+ Import count feature implemented for gene counts

+ new subindex feature and automatic graph creation
(new `wBuild` version, but not yet working on pip, currently github version is used for testing)
+ tests implemented

### CODE REVIEW

File: drop/modules/aberrant-expression-pipeline/Counting/Summary.R [hunk 1]
```diff
@@ -42,7 +41,7 @@ cnts_mtx <- counts(ods, normalized = F)
 #' 
 #' Compare number of records vs. read counts
 #' 
-bam_coverage <- fread(snakemake@input$bam_cov)
+bam_coverage <- fread(snakemake@input$bam_cov, )
 bam_coverage[, sampleID := as.character(sampleID)]
```
Comment (vyepez88) [inline drop/modules/aberrant-expression-pipeline/Counting/Summary.R]: why is there a comma there?

File: drop/modules/aberrant-splicing-pipeline/Counting/01_1_countRNA_splitReads_samplewise.R [hunk 2]
```diff
@@ -2,23 +2,25 @@
...
 #'  threads: 3
 #'  type: script
 #'---
-saveRDS(snakemake, file.path(snakemake@params$tmpdir, "FRASER_01_1.snakemake"))
-# snakemake <- readRDS(".drop/tmp/AS/FRASER_01_1.snakemake")
 
-source("Scripts/_helpers/config.R")
+saveRDS(snakemake, snakemake@log$snakemake)
+source(snakemake@params$setup, echo=FALSE)
+library(BSgenome.Hsapiens.UCSC.hg19)
```
Comment (vyepez88) [inline drop/modules/aberrant-splicing-pipeline/Counting/01_1_countRNA_splitReads_samplewise.R:23]: should be as before

File: drop/modules/mae-pipeline/MAE/Results.R [hunk 3]
```diff
@@ -56,38 +57,30 @@ fo <- findOverlaps(rmae_ranges, gene_annot_ranges)
 # Add the gene names
 res_annot <- cbind(rmae[from(fo), ],  gene_annot_dt[to(fo), .(gene_name, gene_type)])
 
-# Prioritize protein coding genes
+# Prioritze protein coding genes
```
Comment (vyepez88) [inline drop/modules/mae-pipeline/MAE/Results.R]: Prioritize*

File: drop/modules/mae-pipeline/MAE/filterSNVs.sh [hunk 4]
```diff
@@ -21,8 +21,7 @@ samtools=$8
 
 tmp=$(mktemp)
 
-# $bcftools annotate --force -x INFO ${vcf_file} |\
-$bcftools view  $vcf_file | grep -vP "^##INFO=" | awk -F'\t' 'BEGIN {OFS = FS} { if($1 ~ /^[^#]/){ $8 = "." }; print $0 }' |\
+$bcftools annotate --force -x INFO ${vcf_file} |\
```
Comment (vyepez88) [inline drop/modules/mae-pipeline/MAE/filterSNVs.sh]: should be as before

File: drop/modules/mae-pipeline/MAE/Results.R [hunk 5]
```diff
@@ -106,19 +99,15 @@ fwrite(res[MAE_ALT == TRUE], snakemake@output$res_signif,
 #+echo=F
 res[, N := .N, by = ID]
 res[MAE == TRUE, N_MAE := .N, by = ID]
-res[MAE == TRUE & MAE_ALT == FALSE, N_MAE_REF := .N, by = ID]
 res[MAE_ALT == TRUE, N_MAE_ALT := .N, by = ID]
-res[MAE == TRUE & MAE_ALT == FALSE & rare == TRUE, N_MAE_REF_RARE := .N, by = ID]
 res[MAE_ALT == TRUE & rare == TRUE, N_MAE_ALT_RARE := .N, by = ID]
 
-rd <- unique(res[,.(ID, N, N_MAE, N_MAE_REF, N_MAE_ALT, N_MAE_REF_RARE, N_MAE_ALT_RARE)])
+rd <- unique(res[,.(ID, N, N_MAE, N_MAE_ALT, N_MAE_ALT_RARE)])
```
Comment (vyepez88) [inline drop/modules/mae-pipeline/MAE/Results.R]: should be as before

File: drop/modules/mae-pipeline/MAE/Results.R [hunk 6]
```diff
@@ -106,19 +99,15 @@ fwrite(res[MAE_ALT == TRUE], snakemake@output$res_signif,
...
 melt_dt <- melt(rd, id.vars = 'ID')
 melt_dt[variable == 'N', variable := '>10 counts']
 melt_dt[variable == 'N_MAE', variable := '+MAE']
-melt_dt[variable == 'N_MAE_REF', variable := '+MAE for\nREF']
-melt_dt[variable == 'N_MAE_ALT', variable := '+MAE for\nALT']
-melt_dt[variable == 'N_MAE_REF_RARE', variable := '+MAE for REF\n& rare']
-melt_dt[variable == 'N_MAE_ALT_RARE', variable := '+MAE for ALT\n& rare']
+melt_dt[variable == 'N_MAE_ALT', variable := '+MAE for ALT']
```
Comment (vyepez88) [inline drop/modules/mae-pipeline/MAE/Results.R]: should be as before

File: drop/template/config.yaml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-projectTitle: Detection of RNA Outlier Pipeline
+projectTitle: "DROP: Detection of RNA Outlier Pipeline"
```
Comment (vyepez88) [inline drop/template/config.yaml]: Outliers* (add an s)

File: drop/config/DropConfig.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+from drop import utils
+from pathlib import Path
+
+class DropConfig:
+
+    CONFIG_KEYS = [
+        # wbuild keys
+        "projectTitle", "htmlOutputPath", "scriptsPath", "indexWithFolderName", "fileRegex", "readmePath",
+        # global parameters
+        "root", "sampleAnnotation", "geneAnnotation", "genomeAssembly", "scanBamParam", "exportCounts", "tools",
```
Comment (vyepez88) [inline drop/config/DropConfig.py]: remove scanBamParam

File: drop/config/DropConfig.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        config_dict = utils.checkKeys(config_dict, keys=["htmlOutputPath", "root", "sampleAnnotation"], check_files=True)
+        config_dict["geneAnnotation"] = utils.checkKeys(config_dict["geneAnnotation"], keys=None, check_files=True)
+
+        config_dict["indexWithFolderName"] = True
+        config_dict["fileRegex"] = ".*\.R"
+        config_dict["wBuildPath"] = utils.getWBuildPath()
+        
+        setKey = utils.setKey
+        setKey(config_dict, None, "genomeAssembly", "hg19")
+        setKey(config_dict, None, "scanBamParam", "null")
```
Comment (vyepez88) [inline drop/config/DropConfig.py]: remove scanBamParam, also from documentation and config template

File: drop/modules/aberrant-splicing-pipeline/Counting/01_1_countRNA_splitReads_samplewise.R (hunk 2 above)
Comment (mumichae) [reply drop/modules/aberrant-splicing-pipeline/Counting/01_1_countRNA_splitReads_samplewise.R:23]: why? in order to make the script paths safer, I created functions that determine the R config path in the snakemake command above. See snakemake@params@setup to check if it points to "<AS_workdir_in_project_folder>/config.R"

File: drop/modules/mae-pipeline/MAE/filterSNVs.sh (hunk 4 above)
Comment (mumichae) [reply drop/modules/mae-pipeline/MAE/filterSNVs.sh]: 034141b

File: drop/config/DropConfig.py (hunk 9 above)
Comment (mumichae) [reply drop/config/DropConfig.py]: b25ed325371e207317d5f305a84d0a8f81a55cab

File: drop/config/DropConfig.py (hunk 8 above)
Comment (mumichae) [reply drop/config/DropConfig.py]: b25ed325371e207317d5f305a84d0a8f81a55cab

File: drop/template/config.yaml (hunk 7 above)
Comment (mumichae) [reply drop/template/config.yaml]: 1e54358e1700556702943c48906f796c20e5a62e

File: drop/modules/aberrant-expression-pipeline/Counting/Summary.R (hunk 1 above)
Comment (mumichae) [reply drop/modules/aberrant-expression-pipeline/Counting/Summary.R]: 1e54358e1700556702943c48906f796c20e5a62e

File: drop/modules/mae-pipeline/MAE/Results.R (hunk 3 above)
Comment (mumichae) [reply drop/modules/mae-pipeline/MAE/Results.R]: 3453cca00ab761def21917f3a718996079b5b7cb

File: drop/modules/mae-pipeline/MAE/Results.R (hunk 5 above)
Comment (mumichae) [reply drop/modules/mae-pipeline/MAE/Results.R]: 3453cca00ab761def21917f3a718996079b5b7cb

File: drop/modules/mae-pipeline/MAE/Results.R (hunk 6 above)
Comment (mumichae) [reply drop/modules/mae-pipeline/MAE/Results.R]: 3453cca00ab761def21917f3a718996079b5b7cb

File: drop/modules/aberrant-splicing-pipeline/Counting/01_1_countRNA_splitReads_samplewise.R (hunk 2 above)
Comment (vyepez88) [reply drop/modules/aberrant-splicing-pipeline/Counting/01_1_countRNA_splitReads_samplewise.R:23]: I meant the library(BSgenome) should be called inside the if() depending on whether it's hg19 or hg38

**Review by vyepez88 — APPROVED** [review]

File: drop/config/Submodules.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            "maxTestedDimensionProportion"
+        ]
+        self.name = "AberrantExpression"
+        self.rnaIDs = self.sa.subsetGroups(self.groups, assay="RNA")
+        self.extRnaIDs = self.sa.subsetGroups(self.groups, assay="GENE_COUNTS", warn=0, error=0)
+
+    def setDefaultKeys(self, dict_):
+        super().setDefaultKeys(dict_)
+        setKey = utils.setKey
+        setKey(dict_, None, "groups", self.sa.getGroups(assay="RNA"))
```
Comment (c-mertes) [inline drop/config/Submodules.py:46]: It is hard to maintain raw defaults if they are scattered around in the package. 

I would suggest to move them into one yaml file, which should look like the example yaml file where all defaults are set. Then it is also easy to change them without changing the `code`.

File: drop/download_data.sh [hunk 11]
```diff
@@ -3,15 +3,19 @@ set -e
...
+	tar -zxvf resource.tar.gz
+	mv resource Data
+else
+    echo "Data directory already exists, is not updated"
+fi
 
 # prepare data
 cd Data
+cp config_relative_wb1.8.yaml ../config.yaml
 python fix_sample_anno.py
```
Comment (c-mertes) [inline drop/download_data.sh:21]: the demo needs some cleanup. the user should not see a fix_sample_anno.py script. nor any other temporary files.

**Review by c-mertes — APPROVED** [review]
Looks good. 

I would suggest to not use snakemake objects within the scripts but rather have defined variables in the beginning and then use them within the script later.

File: drop/download_data.sh [hunk 12]
```diff
@@ -3,15 +3,19 @@ set -e
 
 # get data
 resource_url="https://www.cmm.in.tum.de/public/paper/drop_analysis/resource.tar.gz"
-wget -Nc $resource_url
-tar -zxvf resource.tar.gz
-rm -rf Data
-mv resource Data
+wget -nc $resource_url
```
Comment (c-mertes) [inline drop/download_data.sh]:
```
tmpresource="$(dirname $(tempfile))/drop_resource.tar.gz"
wget -nc -O$tmpresource $resource_url
```
We should use this so we can download it once and keep it in temp.

**Review by c-mertes — APPROVED** [review]

File: drop/config/Submodules.py (hunk 10 above)
Comment (mumichae) [reply drop/config/Submodules.py:46]: Will address this in a new DROP version

File: drop/download_data.sh (hunk 11 above)
Comment (mumichae) [reply drop/download_data.sh:21]: Will address this in a new DROP version

File: drop/download_data.sh (hunk 12 above)
Comment (mumichae) [reply drop/download_data.sh]: fixed in 9640e3f

**Comment by mumichae** [comment]
Red build currently coming from an incompatibility between `biomaRt` and `R==4.0.3`. For this DROP version I will constrain to `R<=4.0.2`, so this error should be resolved.

**Pull request merged**